2_adventures/001/jbrains/06/langtons_ant.rb

```diff
@@ -0,0 +1,127 @@
+# Read about Langton's Ant at http://link.jbrains.ca/1FGWigr
+#
+# I first saw Kent Beck demonstrate this at XP Universe 2001.
+# To my knowledge, I've never tried writing it myself.
+
+Vector = Struct.new(:x, :y)
+class Location < Vector
+  def step_towards(direction)
+    Location.new(self.x + direction.x, self.y + direction.y)
+  end
+
+  def to_s
+    "(#{x}, #{y})"
+  end
+end
+
+# black_squares: List of black-colored squares
+Grid = Struct.new(:black_squares) 
+class Grid
+  def self.with_black_squares(black_squares)
+    self.new(black_squares)
+  end
+
+  def describe
+    "[#{black_squares.collect(&:to_s).join(", ")}]"
+  end
+
+  def flip_square_at(location)
+    # assume white square
+    black_squares << location
+  end
+end
+
+# CONTRACT Directions are unit Vectors
+class Direction < Vector
+  # how to make this array circular?
+  @@directions_rotating_left = [
+    Direction.new(0, 1),
+    Direction.new(-1, 0),
+    Direction.new(0, -1),
+    Direction.new(1, 0)
+  ]
+
+  def self.north
+    self.new(0, 1)
+  end
+  
+  def self.west
+    self.new(-1, 0)
+  end
+
+  def turn_left
+    # ugh. maintain a cursor, maybe?
+    @@directions_rotating_left[@@directions_rotating_left.index(self) + 1]
+  end
+
+  @@named_directions = {
+    Direction.new(0, 1) => "north",
+    Direction.new(-1, 0) => "west",
+    Direction.new(0, -1) => "south",
+    Direction.new(1, 0) => "east"
+  }
+  def to_s
+    @@named_directions[self]
+  end
+end
+
+LangtonsAnt = Struct.new(:grid, :facing, :location)
+class LangtonsAnt
+  def self.start(args)
+    self.new(args[:grid], args[:facing], args[:location])
+  end
+
+  def go
+    self.grid.flip_square_at(self.location)
+    # assume a white square
+    now_facing = self.facing.turn_left
+    now_at = self.location.step_towards(now_facing)
+    self.class.new(self.grid, now_facing, now_at)
+  end
+end
+
+class WalkListener
+  def step_taken(number, ant)
+    # Optional subclass responsibility
+  end
+end
+
+class ConsoleReportingWalkListener < WalkListener
+  def step_taken(number, ant)
+    if number == 0
+      puts "At the beginning of his walk..."
+    else
+      puts "After step #{number}"
```

Comment:
Yes. One consequence of working without tests: not as much pressure to
separate that behavior. Not yet, anyway. I'm perfectly happy with all the
calls to STDOUT in one place for easy separation when it becomes more
urgent.
On Apr 18, 2015 5:45 PM, "Victor Goff" <notifications@github.com> wrote:

> In 2_adventures/001/jbrains/06/langtons_ant.rb
> <https://github.com/RubySteps/21-day-challenge/pull/361#discussion_r28647289>
> :
>
> > +    self.class.new(self.grid, now_facing, now_at)
> > +  end
> > +end
> > +
> > +class WalkListener
> > +  def step_taken(number, ant)
> > +    # Optional subclass responsibility
> > +  end
> > +end
> > +
> > +class ConsoleReportingWalkListener < WalkListener
> > +  def step_taken(number, ant)
> > +    if number == 0
> > +      puts "At the beginning of his walk..."
> > +    else
> > +      puts "After step #{number}"
>
> Glad you asked. :)
>
> if true
>   puts "This was true"
> else
>   puts "This was false"
> end
>
> result_of_conditional = if true
>                           "This was true"
>                         else
>                           false
>                         end
>
> puts result_of_conditional
>
>
> You are unconditionally doing something in both branches, in this case you
> are using puts unconditionally.
>
> And you are right, you aren't ready for I18n or anything like that where
> you are wanting to take the resulting strings and translate them to another
> language, or wanting the user of your methods to do funky things with the
> method, like reversing the string. So the returned object is nil, because
> you don't yet care what that object is. And that is OK.
>
> But yeah, I would probably use puts when the method is called rather than
> in the method. The puts being either to a logger to log the result, or to
> the screen to present it. or to a database, or to a file Once you do care
> about the return, then the freedom is there.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/RubySteps/21-day-challenge/pull/361/files#r28647289>.
>